homeassistant/components/zha/light.py

```diff
@@ -438,6 +441,28 @@ def async_restore_last_state(self, last_state):
         if "effect" in last_state.attributes:
             self._effect = last_state.attributes["effect"]
 
+    async def async_turn_on(self, **kwargs):
+        """Turn the entity on."""
+        await self._handle_on_off(super().async_turn_on, kwargs)
+
+    async def async_turn_off(self, **kwargs):
+        """Turn the entity off."""
+        await self._handle_on_off(super().async_turn_off, kwargs)
+
+    async def _handle_on_off(self, operation, kwargs):
+        """Perform the on or off operation."""
+        transition = kwargs.get(light.ATTR_TRANSITION)
+        default_transition = DEFAULT_TRANSITION / 10 if DEFAULT_TRANSITION > 1 else 1
+        duration = transition + 0.5 if transition else default_transition + 0.5
+        self._transitioning = True
+        await operation(**kwargs)
+
+        async def do_refresh(_):
+            self._transitioning = False
+            await self.async_get_state()
```

Comment:
If a long transition of 10 seconds for brightness or color is happening, ``do_refresh(_)`` / ``refresh_members(_)`` are called after 10 seconds (as expected). However, when the light is changed to another state (with another long transition) while it is still transitioning, the ``async_call_later`` listener/task (-> do_refresh) will execute while the second transition is happening. This will reset ``_transitioning`` to false (thus parsing attribute reports again) and polls the light "mid transition" (weird temporary color might show up in the Home Assistant UI).
Sometimes, the previous poll seems to set the state in Home Assistant after the latter poll already has done so thus resulting in a wrong, weird color (not sure why the order is messed up though -> maybe because of networking "overloading" when doing this).
(Overall it's much better with the changes though, so thanks!)